pkg/operator/cockroachdb/cluster/controller.go

```diff
@@ -0,0 +1,263 @@
+	}
+}
+
+// Reconcile reads that state of the cluster for a Cluster object and makes changes based on the state read
+// and what is in the Cluster.Spec
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileCockroachDBCluster) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// workaround because the rook logging mechanism is not compatible with the controller-runtime loggin interface
+	reconcileResponse, err := r.reconcile(request)
+	if err != nil {
+		logger.Errorf("failed to reconcile %v", err)
+	}
+
+	return reconcileResponse, err
+}
+
+func (r *ReconcileCockroachDBCluster) reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Fetch the Cluster instance
+	cluster := &cockroachdbv1alpha1.Cluster{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, cluster)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Debug("Cluster resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errors.Wrapf(err, "failed to get Cluster")
+	}
+
+	// The CR was just created, initializing status fields
+	if cluster.Status == nil {
+		cluster.Status = &cockroachdbv1alpha1.Status{}
+		cluster.Status.Phase = k8sutil.Created
+		err := opcontroller.UpdateStatus(r.client, cluster)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+		}
+	}
+
+	// // Set a finalizer so we can do cleanup before the object goes away
+	// err = opcontroller.AddFinalizerIfNotPresent(r.client, cluster)
+	// if err != nil {
+	// 	return reconcile.Result{}, errors.Wrap(err, "failed to add finalizer")
+	// }
+
+	// DELETE: the CR was deleted
+	if !cluster.GetDeletionTimestamp().IsZero() {
+		logger.Debugf("deleting cluster %q", cluster.Name)
+		err := DeleteCluster(r.context, cluster)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrapf(err, "failed to delete cluster %q. ", cluster.Name)
+		}
+
+		// Remove finalizer
+		err = opcontroller.RemoveFinalizer(r.client, cluster)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to remove finalizer")
+		}
+
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
+	}
+
+	// validate the cluster settings
+	if err := ValidateCluster(r.context, cluster); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "invalid cluster CR %q spec", cluster.Name)
+	}
+
+	// Start object reconciliation, updating status for this
+	cluster.Status.Phase = k8sutil.ReconcilingStatus
+	err = opcontroller.UpdateStatus(r.client, cluster)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+	}
+
+	// CREATE/UPDATE
+	reconcileResponse, err := r.reconcileCreateCluster(cluster)
```

Comment:
`controller-runtime` have utility for doing create or update object [controllerutil.CreateOrUpdate](https://github.com/kubernetes-sigs/controller-runtime/blob/a2d55b5721ce735d0f43f955bf83df87db2ce1fe/pkg/controller/controllerutil/controllerutil.go#L192)
is it possible to implement that? although in this PR we have not implemented the update yet, but in my opinion, we will be fine if use that utility.